content/en/blog/_posts/TBD-admission-controllers-for-container-drift/index.md

```diff
@@ -0,0 +1,90 @@
+
+## Admission control for interactive kubectl commands
+Once a request is sent to K8s, it needs to be authenticated and authorized by the API server to proceed. Additionally, K8s has a separate layer of protection called [admission controllers](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/), which can intercept the request before an object is persisted in *etcd*. There are various predefined admission controls compiled into the API server binary (e.g. ResourceQuota to enforce hard resource usage limits per namespace). Besides, there are two dynamic admission controls named [MutatingAdmissionWebhook](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#mutatingadmissionwebhook) and [ValidatingAdmissionWebhook](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#validatingadmissionwebhook), used for mutating or validating K8s requests respectively. The latter is what we adopted to detect container drift at runtime caused by interactive kubectl commands. This whole process can be divided into three steps as explained in detail below.
+
+### 1. Admit interactive kubectl command requests
+First of all, we need to enable a validating webhook that sends qualified requests to our admission controller service. In the interest of interactive kubectl commands, we configure the webhook’s rules with resources as "[pods/exec, pods/attach]", and operations as “CONNECT”. They tell the K8s API server that all “exec” and "attach" requests should be subject to our admission controller service. We also specify a "service" reference (or "url" that gives the location of the webhook) and “caBundle” to provide its certificate, both under the “clientConfig” stanza. Here is a short example of what our ValidatingWebhookConfiguration object looks like:
+```yaml
+apiVersion: admissionregistration.k8s.io/v1
+kind: ValidatingWebhookConfiguration
+metadata:
+  name: example-validating-webhook-config
+webhooks:
+  - name: validate-pod-interaction.example.com
+    sideEffects: None
+    rules:
+      - apiGroups: ["*"]
+        apiVersions: ["*"]
+        operations: ["CONNECT"]
+        resources: ["pods/exec", "pods/attach"]
+    failurePolicy: Fail
+    clientConfig:
+      service:
+        name: example-service
+        namespace: kube-exec-controller
+        path: "/admit-pod-interaction"
+      caBundle: "{{CABUNDLE_VALUE}}"
+    admissionReviewVersions: ["v1", "v1beta1"]
+```
+
+### 2. Tag the pod with potentially mutated containers
```

Comment:
Change “Tag” to “Annotate” and /r or “Label” depending on which is more appropriate.
For example:
```
### 2. Label and annotate the target pod
```

Reply:
Updated to `2. Label the target Pod with potentially mutated containers`